crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -0,0 +1,578 @@
+                                            // notify_service.add_watcher(
+                                            //          "/path/to/some/place",
+                                            //          None,            <------ note the file name is not given
+                                            //          &[Masks::Created]
+                                            //  )
+                                            // here the file we are watching is not known to us, so we match only on event mask
+                                            if event_mask.eq(flag) {
+                                                let full_path = dir_path.join(notify_file_name);
+                                                yield (full_path, event_mask)
+                                            }
+                                        }
+                                    }
+                                }
+
+                            }
+                        }
+                        // there should never be an "if let None = &event.name" because add_watcher
+                        // will always add a watcher as a directory
+                    },
+                    Err(error) => {
+                        // any error comming out of `notify_service.next()` will be
+                        // an std::Io error: https://docs.rs/inotify/latest/src/inotify/stream.rs.html#48
+                        yield Err(NotifyStreamError::FromIOError(error))?;
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// utility function to return an fs notify stream:
+///
+/// this supports both file wathes and directory watches:
+///
+/// # Example
+/// ```rust
+/// use tedge_utils::fs_notify::{fs_notify_stream, FileEvent};
+/// use tedge_test_utils::fs::TempTedgeDir;
+///
+/// // created a new tmp directory with some files and directories
+/// let ttd = TempTedgeDir::new();
+/// let file_a = ttd.file("file_a");
+/// let file_b = ttd.file("file_b");
+/// let file_c = ttd.dir("some_directory").file("file_c");
+///
+///
+/// let fs_notification_stream = fs_notify_stream(&[
+///      (ttd.path(), Some(String::from("file_a")), &[FileEvent::Created]),
+///      (ttd.path(), Some(String::from("file_b")), &[FileEvent::Modified, FileEvent::Created]),
+///      (ttd.path(), Some(String::from("some_directory/file_c")), &[FileEvent::Deleted])
+///     ]
+/// ).unwrap();
+/// ```
+pub fn fs_notify_stream(
+    input: &[(&Path, Option<String>, &[FileEvent])],
+) -> Result<impl Stream<Item = Result<(PathBuf, FileEvent), NotifyStreamError>>, NotifyStreamError>
+{
+    let mut fs_notification_service = NotifyStream::try_default()?;
+    for (dir_path, watch, flags) in input {
+        fs_notification_service.add_watcher(dir_path, watch.to_owned(), flags)?;
+    }
+    Ok(fs_notification_service.stream())
+}
+
+#[cfg(test)]
+#[cfg(feature = "fs-notify")]
```

Comment:
Just `cargo run` runs these tests or `feature` flag has to be added with `cargo run`?. If its later, workflow update is needed, and also I feel it's not a good idea because we might hide these tests under some feature flag.

Reply:
Oh you're right, flag needs to be added, working on it